roxie/ccd/ccdfile.cpp

```diff
@@ -1270,6 +1616,211 @@ class CRoxieFileCache : implements IRoxieFileCache, implements ICopyFileProgress
         }
     }
 
+    static unsigned __int64 readPage(const char * &_t)
+    {
+        const char *t = _t;
+        unsigned __int64 v = 0;
+        for (;;)
+        {
+            char c = *t;
+            if ((c >= '0') && (c <= '9'))
+                v = v * 16 + (c-'0');
+            else if ((c >= 'a') && (c <= 'f'))
+                v = v * 16 + (c-'a'+10);
+            else if ((c >= 'A') && (c <= 'F'))
+                v = v * 16 + (c-'A'+10);
+            else
+                break;
+            t++;
+        }
+        _t = t;
+        return v;
+    }
+
+    virtual void loadSavedOsCacheInfo() override
+    {
+        Owned<const ITopologyServer> topology = getTopology();
+        for (unsigned channel : topology->queryChannels())
+            doLoadSavedOsCacheInfo(channel);
+        doLoadSavedOsCacheInfo(0);  // MORE - maybe only if I am also a server?
+    }
+
+    void doLoadSavedOsCacheInfo(unsigned channel)
+    {
+        const char* dllserver_root = getenv("HPCC_DLLSERVER_PATH");
+        assertex(dllserver_root != nullptr);
+        VStringBuffer cacheFileName("%s/%s/cacheInfo.%d", dllserver_root, roxieName.str(), channel);
+        StringBuffer cacheInfo;
+        try
+        {
+            if (checkFileExists(cacheFileName))
+            {
+                cacheInfo.loadFile(cacheFileName, false);
+                warmOsCache(cacheInfo);
+                if (traceLevel)
+                    DBGLOG("Loaded cache information from %s for channel %d", cacheFileName.str(), channel);
+            }
+        }
+        catch(IException *E)
+        {
+            EXCLOG(E);
+            E->Release();
+        }
+    }
+
+
+    virtual void warmOsCache(const char *cacheInfo) override
+    {
+        if (!cacheInfo)
+            return;
+        size_t os_page_size = getpagesize();
+        char t = 0;
+        unsigned touched = 0;
+        unsigned preloaded = 0;
+        Owned<const ITopologyServer> topology = getTopology();
+        while (*cacheInfo)
+        {
+            // MORE - what if there is a | in the filename?
+            unsigned fileChannel = strtoul(cacheInfo, (char **) &cacheInfo, 10);
+            if (*cacheInfo != '|')
+                break;
+            if (!topology->implementsChannel(fileChannel))
+            {
+                const char *eol = strchr(cacheInfo, '\n');
+                if (!eol)
+                    break;
+                cacheInfo = eol+1;
+                continue;
+            }
+            cacheInfo++;
+            const char *endName = strchr(cacheInfo, '|');
+            assert(endName);
+            if (!endName)
+                break;
```

Comment:
is this unexecpted, should it throw an error or warning? (and similar on line 1686)

Reply:
It will throw an assert in debug mode, but ignore invalid data in release. That was deliberate.